mipengine/node/monetdb_interface/guard.py

```diff
@@ -0,0 +1,229 @@
+
+    Examples
+    --------
+    >>> @sql_injection_guard(a=str.isalnum, b=my_validator, c=None)
+    ... def f(a, b, c):
+    ...     ...
+    """
+
+    def decorator(func):
+        argspec = inspect.getfullargspec(func)
+        if set(argspec.args) != set(validators.keys()):
+            raise ValueError(
+                "sql_injection_guard validators do not match function "
+                f"{func.__name__} args.\n"
+                f"Args mismatch: {set(argspec.args) - set(validators.keys())}.\n"
+                "Make sure to add one validator per function arg, passing "
+                "None for args that don't need validation."
+            )
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            values = get_arg_values(argspec, args, kwargs)
+            validate_arg_values(validators, values)
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def get_arg_values(argspec: FullArgSpec, args: list, kwargs: dict) -> dict:
+    arg_values = get_named_defaults(argspec)
+    arg_values.update(get_named_posarg_values(argspec, args))
+    arg_values.update(kwargs)
+    return arg_values
+
+
+def get_named_defaults(argspec: FullArgSpec) -> dict:
+    if argspec.defaults:
+        return dict(zip(reversed(argspec.args), reversed(argspec.defaults)))
+    return {}
+
+
+def get_named_posarg_values(argspec: FullArgSpec, args: list) -> dict:
+    return dict(zip(argspec.args, args))
+
+
+def validate_arg_values(validators: dict, arg_values: dict) -> None:
+    for arg, value in arg_values.items():
+        validator = validators[arg] or (lambda x: True)
+        if not validator(value):
+            raise InvalidSQLParameter(
+                f"Arg '{arg}' with value '{value}' cannot be validated "
+                f"with validator '{validator.__name__}'."
+            )
+
+
+class InvalidSQLParameter(Exception):
+    pass
+
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Validators ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# ip regex
+#   https://stackoverflow.com/questions/5284147/validating-ipv4-addresses-with-regexp
+# hostname regex (RFC 952)
+#   https://stackoverflow.com/questions/106179/regular-expression-to-match-dns-hostname-or-ip-address
+ip_re = r"((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}"
+alpha_re = r"[a-zA-Z]"
+alnum_re = r"[a-zA-Z0-9]"
+alnumhyphen_re = r"[a-zA-Z0-9-]"
+hostname_re = rf"""(({alpha_re}|{alpha_re}{alnumhyphen_re}*{alnum_re})\.)*
+                    ({alpha_re}|{alpha_re}{alnumhyphen_re}*{alnum_re})"""
+port_re = r"(?P<port>\d{1,5})"
+socketaddress_ptrn = re.compile(rf"(?:{ip_re}|{hostname_re}):{port_re}", re.A | re.X)
+version_re = r"\w+(\.\w)*"
+datamodel_re = rf"\w+:{version_re}"
+datamodel_ptrn = re.compile(datamodel_re, re.A)
+sqlidentifier_re = r"[a-z_][a-z0-9_]*"
+datatable_ptrn = re.compile(rf'"{datamodel_re}"\.("\w+"|{sqlidentifier_re})', re.A)
+
+
+def is_lowercase_identifier(string):
+    return str.isidentifier(string) and str.islower(string)
```

Comment:
Why use `is_lowercase_identifier` instead of the `isidentifier`?

We are testing for sql injection here and there is no need to enforce something different in this context.

Reply:
Yes that's true. However, I really wanted to remove [this line](https://github.com/madgik/MIP-Engine/blob/9a275b047229226f4a6e05299a5bae4d8e18d191/mipengine/node/tasks/udfs.py#L461). 

The thing is, we could have one set of validators for SQL injections and another set of validators for verifying the validity of SQL statements. However, they have a lot of overlap so I think it's more simple to merge them. 

We could change the name `sql_injection_guard` to simply `sql_guard` or `validate_sql` or something. What do you think?

Reply:
I am not sure I follow. How could this line https://github.com/madgik/MIP-Engine/blob/9a275b047229226f4a6e05299a5bae4d8e18d191/mipengine/node/tasks/udfs.py#L461 be removed by checking for lowercase characters?